i#5505 kernel memtrace: Add submodule libipt & Initialize submodules all the time

dolanzhao opened the pull request:
- Add libipt as a submodule
- Tell GA CI to init submodules: submodules: true arg to checkout step
- Add submodule update to make/git/devsetup.sh
- Add submodule update to make/git/git_pullall.sh
- Update how-to-build docs

Issue: #5505

derekbruening commented:
> [#5515 (comment)](https://github.com/DynamoRIO/dynamorio/pull/5515#discussion_r889369065)
> 
> > Potentially add a pre-commit hook to prevent the common mistake of accidentally rolling back a recent submodule update when switching branches
> 
> @derekbruening Could you explain this point in details?

If you check out the drmemory repo and do "git grep submodule" you will see all the support it has: it is part of make/git/hook-pre-commit.sh that looks useful for the DR repo.

derekbruening commented on api/docs/building.dox:
```diff
@@ -47,7 +47,7 @@ To build DynamoRIO on Linux, use the following commands as a guide.  This builds
   $ sudo apt-get install cmake g++ g++-multilib doxygen git zlib1g-dev libunwind-dev libsnappy-dev liblz4-dev
   # Get sources.
-  $ git clone https://github.com/DynamoRIO/dynamorio.git
+  $ git clone --recurse https://github.com/DynamoRIO/dynamorio.git
```
nit: I think the actual option is either --recursive or --recurse-submodules?  I guess it is allowing your abbreviation but maybe better to use the real name.

dolanzhao replied:
Done.

derekbruening commented on .github/workflows/ci-package.yml:
```diff
@@ -59,6 +59,8 @@ jobs:
     steps:
     - uses: actions/checkout@v2
+      with:
+        submodules: true
```
The checkout step is used in all the other .yml files too

dolanzhao replied:
Done.

derekbruening commented on make/git/git_pullall.sh:
```diff
@@ -44,6 +44,8 @@ if test -z "${has_remote}"; then
 else
     echo "First, updating with rebase from remote ${branch}."
     git pull --rebase origin ${branch}
+    echo "\nSecond, updating submodules."
```
This will print a literal `\n`.  Use -e to interpret newlines.

dolanzhao replied:
Done.

derekbruening commented on make/git/git_pullall.sh:
```diff
@@ -44,6 +44,8 @@ if test -z "${has_remote}"; then
 else
```
The submodule update should be in the if case above too

dolanzhao replied:
Done.

derekbruening commented on make/git/git_pullall.sh:
```diff
@@ -44,6 +44,8 @@ if test -z "${has_remote}"; then
     echo "First, updating with rebase from remote ${branch}."
     git pull --rebase origin ${branch}
+    echo "\nSecond, updating submodules."
+    git submodule update --init --recursive
```
Hmm, drmemory does this after the pull below; I think that is what we want, to reflect the remote main branch changes

dolanzhao replied:
Done.

derekbruening commented on .gitmodules:
```diff
@@ -0,0 +1,3 @@
+[submodule "third_party/libipt"]
+    path = third_party/libipt
+    url = https://github.com/intel/libipt.git
```
I had assumed that libipt was not provided in the main distro repos: but it does seem to be there as `libipt-dev`.  It would be simpler to use that package, just like we do with many other libraries (e.g., for dr$sim compression: libsnappy, lz4, zlib).  Is there some very recent interface that is not part of that package that is the reason to build it from sources?

dolanzhao replied:
Yes. libipt-dev only contains libipt.so. It can not decode perf recode trace data directly. Currently, drptpostprocessor uses the libipt sideband library (libipt-sb.so) to help handle context switches across images. So we need to build it from the source file. drptpostprocessor is a general solution for perf record data decoder. Using the sideband APIs is a quick solution because we don't need to write more code to handle sideband messages.
But for the data generated by our recorder, the kernel trace instructions may all be decoded from kcore, and we may only need libipt.so. In this case, libipt-dev might meet our requirements.

dolanzhao replied:
I discussed this question with @abhinav92003 this morning. We think for the current version, there are three reasons that we need to add libipt as a submodule.
- The first reason is what I mentioned above, we need drptpostprocessor to handle trace data recorded by perf, using libipt-sb.so's API is an easy solution. 
- The second reason is except for the image switch, we may need to handle other information like timestamps in the future. 
- The last reason is we may need to handle the image switch when decoding kernel traces.  
So do you agree that we can include it first and if we don't need it we can remove it in the future.

derekbruening replied:
OK, seems reasonable.  This is a good reason to isolate all the submodule parts in a separate PR so it can be reverted on its own.

derekbruening commented on .github/workflows/ci-aarchxx.yml:
```diff
@@ -55,6 +55,8 @@ jobs:
 
```
The PR title:

> i#5517 add submodule libipt: Add libipt & Initialize submodules all the time #5518

The convention is to only use the short issue description before the colon for issue numbers addresses with multiple PR's.  Since you've made this single issue just for this submodule, with presumably a single PR addressing it, we would expect to not have that before-colon issue description.  (Or, put this PR under the #5505 issue and close #5517 as a duplicate/part of #5505.  If not doing that: please add a ref to #5505 in the Issue: list in the commit description.)

dolanzhao replied:
Done.

derekbruening commented on make/git/hook-pre-commit.sh:
```diff
@@ -57,4 +57,30 @@ if [ -z "${existing}" ]; then
+# submodule + other stuff and if so bail if the submodule version is older.
+# We want to allow a submodule-only rollback.
+# XXX: this relies on several unix utilities which we assume are available
+# on Windows.
```
Looks like a close copy from the drmemory version: but there's no really simple way to share unfortunately.  Maybe add a XXX comment that this is mostly a verbatim copy and if we set up some way to share (dedicated separate repository that's a submodule in both that provides shared config files or sthg?) we should be sure to share this too.

dolanzhao replied:
Done.

abhinav92003 commented:
Thanks for addressing the comments! In addition to commenting "done" in each conversation (which you did), please also click on the "resolve conversation" so that it doesn't show up anymore.

dolanzhao commented:
> Thanks for addressing the comments! In addition to commenting "done" in each conversation (which you did), please also click on the "resolve conversation" button so that it doesn't show up anymore.

OK. Thanks!

dolanzhao commented on api/docs/building.dox:
```diff
@@ -47,7 +47,7 @@ To build DynamoRIO on Linux, use the following commands as a guide.  This builds
   $ sudo apt-get install cmake g++ g++-multilib doxygen git zlib1g-dev libunwind-dev libsnappy-dev liblz4-dev
   # Get sources.
-  $ git clone https://github.com/DynamoRIO/dynamorio.git
+  $ git clone --recursive https://github.com/DynamoRIO/dynamorio.git
```
Done.

abhinav92003 commented:
run arm tests
